modules/order/src/AvailabilityOrderProcessor.php

```diff
@@ -37,10 +38,23 @@ public function process(OrderInterface $order) {
     foreach ($order->getItems() as $order_item) {
       $purchased_entity = $order_item->getPurchasedEntity();
       if ($purchased_entity) {
-        $available = $this->availabilityManager->check($purchased_entity, $order_item->getQuantity(), $context);
-        if (!$available) {
-          $order->removeItem($order_item);
-          $order_item->delete();
+        /** @var \Drupal\commerce\AvailabilityResponseInterface $availability */
+        $availability = $this->availabilityManager->check($purchased_entity, $order_item->getQuantity(), $context);
+        if (!$availability->isNeutral()) {
+          if ($availability->getMax() == 0) {
+            $order->removeItem($order_item);
+            $order_item->delete();
+            drupal_set_message(t('The item %item is no longer available and has been removed from your cart.', [
+              '%item' => $order_item->getTitle(),
+            ]), 'warning');
+          }
+          elseif ($availability->getMax() < $order_item->getQuantity()) {
+            drupal_set_message(new PluralTranslatableMarkup(abs($availability->getMax() - $order_item->getQuantity()), '1 %item is currently unavailable. The quantity in your cart has been updated.', '@count %item are currently unavailable. The quantity in your cart has been updated.', [
+              '%item' => $order_item->getTitle(),
+            ]), 'warning');
+            $order_item->setQuantity($availability->getMax());
+            $order_item->save();
+          }
```

Comment:
I think I get why there is no check for ->isAvailable here, but I will walk through how this would work for the custom backorder availability checker I think I need to implement. 
- my checker should run first, and in the case that the item has 0 stock but can be backordered, we return an Availability response with isAvailable == true, and we might think that is all we need to do, but..
- that isAvailable TRUE doesn't tell us anything about what amount of stock it means to be available, so even though our checker knows that it means "any amount" we have to explicitly answer with a large getMax in this case. 
- the edge condition here, for me, is a order item that has positive stock (deliverable immediately) AND can be backordered (N-time delivery).  If, for instance, stock was 2, then 2 items were put in cart A, but not checked out, then someone else added 1 item to cart B and checked out. Cart A refresh will still (correctly) get an isAvailable response here, with a very high max. However by default there is no way to tell them that now since stock is 1 not 2, one of the items won't deliver immediately anymore. That is probably the task of a custom OrderProcessor just for updating e.g. a custom OrderItem delivery_time field and maybe dsm the customer.

Reply:
@ransomweaver Are you still working on/using backorderable? I'd like to know how this PR affects your use cases.

Reply:
@steveoliver Yes, I'll check this out and see what it does for backorderable.

Reply:
@steveoliver commerce_stock_backorderable definitely depends on this PR. Its checker returns Drupal\commerce\AvailabilityResponse rather than a boolean.  After some tweaks, everything about backorderable seems to work properly with the latest version of this PR. 
In another comment here I've noted that I have a separate issue depending on this PR (https://www.drupal.org/node/2847529) having nothing to do with backorderable. This add-to-cart form availability issue I have chosen to solve with a hook_form_alter in commerce_stock_local, as that seems the best place for it. I will turn that code into a PR for commerce_stock.